neqo-http3/src/push_stream.rs

```diff
@@ -0,0 +1,162 @@
+            _ => None,
+        }
+    }
+}
+
+// The `PushController` keeps information about all push streams. Each push stream is responsible for contacting the
+// `PushController` to consult it about the push state and to inform it when the push stream is done (this are signal
+// from the peer: stream has been closed or reset). `PushController` handles CANCEL_PUSH frames and canceling
+// push from applications and PUSH_PROMISE frames.
+//
+// `PushStream` is responsible for reading from a push stream. It is used for reading push_id as well.
+// It is created when a new push stream is received.
+//
+// After push_id has been read, The `PushController` is informed about the stream and its push_id. This is done by
+// calling `add_new_push_stream`. `add_new_push_stream` may return an error (this will be a connection error)
+// or a bool. true means that the streams should continue and false means that the stream should be reset(the stream
+// will be canceled if the push has been canceled already (CANCEL_PUSH frame or canceling push from the application)
+//
+// `PushStreams` are kept in Http3Connection::recv_streams the same as a normal request/response stream.
+// Http3Connection and read_data is responsible for reading the push data.
+//
+// PushHeaderReady and PushDataReadable are posted through the `PushController` that may decide to postpone them if
+// a push_promise has not been received for the stream.
+//
+// `PushStream` is responsible for removing itself from the `PushController`.
+//
+// `PushStream` may be reset from the peer in the same way as a request stream. The `PushStream` informs the
+// `PushController` that will set the push state to closed and remove any push events.
+
+#[derive(Debug)]
+pub(crate) struct PushStream {
+    state: PushStreamState,
+    stream_id: u64,
+    push_handler: Rc<RefCell<PushController>>,
+    events: Http3ClientEvents,
+}
+
+impl PushStream {
+    pub fn new(
+        stream_id: u64,
+        push_handler: Rc<RefCell<PushController>>,
+        events: Http3ClientEvents,
+    ) -> Self {
+        Self {
+            state: PushStreamState::ReadPushId(NewStreamTypeReader::new()),
+            stream_id,
+            push_handler,
+            events,
+        }
+    }
+}
+
+impl Display for PushStream {
+    fn fmt(&self, f: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {
+        write!(f, "Push stream {:?}", self.stream_id)
+    }
+}
+
+impl RecvStream for PushStream {
+    fn receive(&mut self, conn: &mut Connection, decoder: &mut QPackDecoder) -> Res<()> {
+        loop {
+            match &mut self.state {
+                PushStreamState::ReadPushId(id_reader) => {
+                    let push_id = id_reader.get_type(conn, self.stream_id);
+                    let fin = id_reader.fin();
+                    if fin {
+                        self.state = PushStreamState::Closed;
+                        return Ok(());
```

Comment:
This really isn't OK, is it?  Is there an event that we need to raise in this case?  It seems like this might be cause for a PushError event.